neurons/miner.py

```diff
@@ -368,51 +373,113 @@ async def get_data_entity_bucket_priority(
         return self.default_priority(synapse)
 
+    async def get_contents_by_buckets(
+        self, synapse: GetContentsByBuckets
+    ) -> GetContentsByBuckets:
+        """Used to bulk expose raw contents for all entities within the requested buckets to validators for user queries."""
+        bt.logging.info(
+            f"Got to a GetContentsByBuckets request from {synapse.dendrite.hotkey} for Bucket IDs: {str(synapse.data_entity_bucket_ids)}."
+        )
+
+        # Get a dict of all the contents by DataEntityBucketId for the requested Buckets.
+        synapse.bucket_ids_to_contents = (
+            self.storage.list_contents_in_data_entity_buckets(
+                synapse.data_entity_bucket_ids
+            )
+        )
+        synapse.version = constants.PROTOCOL_VERSION
+
+        bt.logging.success(
+            f"Returning Bucket IDs: {str(synapse.data_entity_bucket_ids)} with {sum(len(v) for v in synapse.bucket_ids_to_contents.values())} entities to {synapse.dendrite.hotkey}."
+        )
+
+        return synapse
+
+    async def get_contents_by_buckets_blacklist(
+        self, synapse: GetContentsByBuckets
+    ) -> typing.Tuple[bool, str]:
+        # Check that the maximum number of buckets to be requested at once is respected.
+        if len(synapse.data_entity_bucket_ids) > constants.BULK_BUCKETS_COUNT_LIMIT:
+            return (
+                True,
+                f"Rejecting GetContentsByBuckets request from {synapse.dendrite.hotkey} at {synapse.dendrite.ip} for requesting {len(synapse.data_entity_bucket_ids)} data entity buckets over limit of {constants.BULK_BUCKETS_COUNT_LIMIT}.",
+            )
+
+        # Check that none of the requested buckets are before the content obfuscation time start.
+        minimum_time_bucket = TimeBucket.from_datetime(
+            constants.REDUCED_CONTENT_DATETIME_GRANULARITY_THRESHOLD
+        )
+        for id in synapse.data_entity_bucket_ids.keys:
+            if id.time_bucket.id < minimum_time_bucket.id:
+                return (
+                    True,
+                    f"Rejecting GetContentsByBuckets request from {synapse.dendrite.hotkey} at {synapse.dendrite.ip} for requesting a data entity bucket: {id} before {constants.REDUCED_CONTENT_DATETIME_GRANULARITY_THRESHOLD}.",
+                )
+        return self.default_blacklist(synapse)
+
+    async def get_contents_by_buckets_priority(
+        self, synapse: GetContentsByBuckets
+    ) -> float:
+        return self.default_priority(synapse)
+
     def default_blacklist(self, synapse: bt.Synapse) -> typing.Tuple[bool, str]:
         """The default blacklist that only allows requests from validators."""
-        if synapse.dendrite.hotkey in [
-            "5Gpt8XWFTXmKrRF1qaxcBQLvnPLpKi6Pt2XC4vVQR7gqNKtU"
-        ]:
+        hotkey = synapse.dendrite.hotkey
+        ip = synapse.dendrite.ip
+        synapse_type = type(synapse)
+
+        if hotkey in ["5Gpt8XWFTXmKrRF1qaxcBQLvnPLpKi6Pt2XC4vVQR7gqNKtU"]:
```

Comment:
Note for future: We should remove this

Reply:
True, went ahead and did that now with the other blacklist changes.